librz/io/p/io_srec.c

```diff
@@ -0,0 +1,574 @@
+// SPDX-FileCopyrightText: 2021 Manolache183 <manolache.alexandru8@gmail.com>
+// SPDX-FileCopyrightText: 2021 swym  <0xfd000000@gmail.com>
+// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+/* Motorola S-record file format
+ * +--------+------------+---------+------+----------+----+
+ * | S type | byte count | address | data | checksum | \n |
+ * +--------+------------+---------+------+----------+----+
+ * 0        1            2         4      N         N+1  N+2
+ *
+ * https://en.wikipedia.org/wiki/SREC_(file_format)
+ */
+
+#include <rz_io.h>
+#include <rz_lib.h>
+#include <rz_util.h>
+
+#define SREC_SIZE        64
+#define NMAX             100
+#define SREC_PATH_PREFIX "srec://"
+
+typedef struct srec_t {
+	st32 fd;
+	RzBuffer *buf;
+} SRecord;
+
+static void write_S3_record(FILE *fd, ut32 address, const ut8 *buffer, ut16 size) {
+	ut8 record_size = 4 + size + 1;
+
+	ut8 checksum = record_size;
+	checksum += address & 0xff;
+	checksum += (address >> 8) & 0xff;
+	checksum += (address >> 16) & 0xff;
+	checksum += address >> 24;
+
+	fprintf(fd, "S3%02x%08x", record_size, address);
+
+	for (ut16 j = 0; j < size; j++) {
+		checksum += buffer[j];
+		fprintf(fd, "%02x", buffer[j]);
+	}
+
+	fprintf(fd, "%02x\n", checksum);
+}
+
+static st32 __write(RzIO *io, RzIODesc *fd, const ut8 *buf, st32 count) {
+	const char *pathname = NULL;
+	FILE *out = NULL;
+	SRecord *srec = NULL;
+	ut32 address = 0, size = 0;
+	ut8 n_bytes = 0;
+	size_t chunks_count = 0;
+	const RzBufferSparseChunk *chunks = NULL;
+	const RzBufferSparseChunk *sparse = NULL;
+
+	if (!fd || !fd->data || (fd->perm & RZ_PERM_W) == 0 || count <= 0) {
+		return -1;
+	}
+
+	srec = (SRecord *)fd->data;
+	pathname = fd->name + strlen(SREC_PATH_PREFIX);
+	out = rz_sys_fopen(pathname, "w");
+
+	// starting record, contains "rizin-srec\0"
+	fprintf(out, "S00E000072697A696E2D7372656300EB\n");
+
+	if (!out) {
+		RZ_LOG_ERROR("srec:write(): cannot open '%s' for writing\n", pathname);
+		return -1;
+	}
+
+	/* mem write */
+	if (rz_buf_write_at(srec->buf, io->off, buf, count) != count) {
+		RZ_LOG_ERROR("srec:write(): cannot write into buffer\n");
+		fclose(out);
+		return -1;
+	}
+
+	rz_buf_seek(srec->buf, count, RZ_BUF_CUR);
+	chunks = rz_buf_sparse_get_chunks(srec->buf, &chunks_count);
+
+	for (size_t i = 0; i < chunks_count; i++) {
+		sparse = &chunks[i];
+		size = sparse->to - sparse->from;
+		for (ut32 offset = 0; offset < size; offset += SREC_SIZE) {
+			address = sparse->from + offset;
+			n_bytes = SREC_SIZE;
+			if (offset + SREC_SIZE > size) {
+				n_bytes = size - offset;
+			}
+			write_S3_record(out, address, sparse->data + offset, n_bytes);
+		}
+	}
+
+	// termination
+	fprintf(out, "S70500000000FA\n");
+	fclose(out);
+	out = NULL;
+	return 0;
+}
+
+static st32 __read(RzIO *io, RzIODesc *fd, ut8 *buf, st32 count) {
+	if (!fd || !fd->data || (count <= 0)) {
```

Comment:
Please also add `rz_return_val_if_fail(io && fd && buf & !count, -1)`. There is no point to allow reads of `0` bytes.